src/zmalloc.c

```diff
@@ -87,10 +87,22 @@ void zlibc_free(void *ptr) {
 #define dallocx(ptr,flags) je_dallocx(ptr,flags)
 #endif
 
-#define update_zmalloc_stat_alloc(__n) atomicIncr(used_memory,(__n))
-#define update_zmalloc_stat_free(__n) atomicDecr(used_memory,(__n))
+#define update_zmalloc_stat_alloc(__n) used_memory_tls[thread_index] += (__n)
+#define update_zmalloc_stat_free(__n) used_memory_tls[thread_index] -= (__n)
 
-static serverAtomic size_t used_memory = 0;
+/* A thread-local storage which keep the current thread's index in the
+ * used_memory_tls array. */
+static __thread int thread_index;
+/* MAX_THREADS_NUM = IO_THREADS_MAX_NUM(128) + BIO threads(3) + main thread(1). */
+#define MAX_THREADS_NUM 132
+static size_t used_memory_tls[MAX_THREADS_NUM];
+
+static serverAtomic int total_active_threads;
+
+/* Register the thread index in start_routine. */
+void zmalloc_register_thread_index(void) {
+    atomicGetIncr(total_active_threads, thread_index, 1);
```

Comment:
```suggestion
    serverAssert(total_active_threads < MAX_THREADS_NUM);
    atomicGetIncr(total_active_threads, thread_index, 1);
```

Reply:
Since both `total_active_threads` and `thread_index` are of `size_t`, whose sizes match the native word size of the platform, the proposed change should work. 

Strictly speaking though (if `size_t` was 64 bit on a 32-bit platform), L112 and L113 need to be flipped. We also shouldn't touch the global var `total_active_threads` without the atomic operation; instead, we should access the thread local var `thread_index`. 

```suggestion
    atomicGetIncr(total_active_threads, thread_index, 1);
    serverAssert(thread_index < MAX_THREADS_NUM);
```